cli/azd/cmd/middleware/environment.go

```diff
@@ -0,0 +1,82 @@
+package middleware
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/azure/azure-dev/cli/azd/cmd/actions"
+	"github.com/azure/azure-dev/cli/azd/internal"
+	"github.com/azure/azure-dev/cli/azd/pkg/environment"
+	"github.com/azure/azure-dev/cli/azd/pkg/environment/azdcontext"
+	"github.com/azure/azure-dev/cli/azd/pkg/lazy"
+)
+
+// EnvironmentMiddleware is a middleware that loads the environment when not readily available
+type EnvironmentMiddleware struct {
+	lazyAzdContext *lazy.Lazy[*azdcontext.AzdContext]
+	lazyEnvManager *lazy.Lazy[environment.Manager]
+	lazyEnv        *lazy.Lazy[*environment.Environment]
+	envFlags       internal.EnvFlag
+}
+
+// NewEnvironmentMiddleware creates a new instance of the EnvironmentMiddleware
+func NewEnvironmentMiddleware(
+	lazyAzdContext *lazy.Lazy[*azdcontext.AzdContext],
+	lazyEnvManager *lazy.Lazy[environment.Manager],
+	lazyEnv *lazy.Lazy[*environment.Environment],
+	envFlags internal.EnvFlag,
+) Middleware {
+	return &EnvironmentMiddleware{
+		lazyAzdContext: lazyAzdContext,
+		lazyEnvManager: lazyEnvManager,
+		lazyEnv:        lazyEnv,
+		envFlags:       envFlags,
+	}
+}
+
+// Run runs the EnvironmentMiddleware to load the environment when not readily available
+func (m *EnvironmentMiddleware) Run(ctx context.Context, next NextFn) (*actions.ActionResult, error) {
+	// We already have an environment, skip loading
+	// This will typically be the case when an environment has been created from a previous command like `azd init`
+	env, err := m.lazyEnv.GetValue()
+	if err == nil && env != nil {
+		return next(ctx)
+	}
+
+	// Needs Azd context before we can have an environment
+	azdContext, err := m.lazyAzdContext.GetValue()
```

Comment:
It's been quite a while since I haven't seen any actual changes to the `azdContext`.

If we look at it, the azdContext seems like a wrapper abstraction on top of a `project directory` with a set of convenience methods:

![image](https://github.com/Azure/azure-dev/assets/24213737/92c5afe9-741f-4fc5-90f3-1b98bb9e1531)

I guess, if we were trying to document what is an azdContext, it would be something like:

> The abstraction for a path where azd should execute. It can be used to get more information about where azd-files should be expected to be found. For example, where should an `azure.yaml` or the `.azure` folder should be. 

That been said, the `azdContext` is providing information for the project files structure, which works great for keeping things like the environment and the project unaware of what is their expected place within an azd project.  But, OTOH, there's never a case, and most likely there will never be a case, where environments (.azure) and/or project (azure.yaml) are placed in different locations within a folder.

We recently had to implement one scenario for Aspire where azure.yaml is not placed at the root of a repo; and even then, we did not update the wat azdContext works. Instead, we only adapted `azd pipeline config` to look for the root of the repo to find the pipeline definitions, but the azdContext is still a static reference.

The reason why I mentioned all of this is because I think azdContext is the real root issue, and we have an opportunity here to design some improvements around azdContext-Environment-Project.

Reply:
There aren't any changes to `AzdContext` for this PR but agree with you that would benefit from refactoring.  Many of the functions here are based on the Project and likely would be a better fit on the `ProjectConfig` itself rather than this component. Let's save this for another day.